### PR TITLE
High availability tasks

### PR DESCRIPTION
**Pull request by reebhub**
Related issues:
[RDoc-2272](https://issues.hibernatingrhinos.com/issue/RDoc-2272/Update-the-High-Availability-Tasks-page-and-related-issues)
[RDoc-2569](https://issues.hibernatingrhinos.com/issue/RDoc-2569/Document-ETL-failover-for-artificial-documents-causing-us-to-miss-documents)
[RDoc-2274](https://issues.hibernatingrhinos.com/issue/RDoc-2274/Pin-tasks-to-specific-node)
[RDoc-2587](https://issues.hibernatingrhinos.com/issue/RDoc-2587/Fix-Samples.csproj-dependencies)

### CODE REVIEW

File: Documentation/5.1/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown [hunk 1]
```diff
@@ -15,13 +15,34 @@
...
 ---
 
+{PANEL: License}
+
+Please [check your license](https://ravendb.net/buy) to verify whether the Highly Available Tasks feature 
+is activated in your database.  
+  
+ * If your license **provides** highly available tasks, the responsibilities of a failed node will be assigned 
+   automatically and immediately to another, available, cluster node.  
+   Reads, Writes, and the ongoing tasks SQL and Raven ETL, Backup, Data subscription, and External replication,
```
[the remark this answers is not in the document]
Comment (karmeli87) [reply Documentation/5.1/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown:33]: Actually I'm not so sure that we enforce this for the replication.. need to check that

File: Documentation/5.2/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown [hunk 2]
```diff
@@ -0,0 +1,145 @@
+# Highly Available Tasks
+---
+
+{NOTE: }
+
+* A **RavenDB Task** can be one of the following:  
+
+  * An [Ongoing Task](../../../studio/database/tasks/ongoing-tasks/general-info)  
+  * Updating a `Rehab` or a `Promotable` [Database Node](../../../server/clustering/distribution/distributed-database#database-topology)
```
Comment (karmeli87) [inline Documentation/5.2/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown:9]: this is more an implementation detail.. not sure it is relevant here, has nothing to do with Highly Available Tasks?

File: Documentation/5.2/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+{PANEL: License}
+
+Please [check your license](https://ravendb.net/buy) to verify whether the Highly Available Tasks feature 
+is activated in your database.  
+  
+ * If your license **provides** highly available tasks, the responsibilities of a failed cluster node will 
+   be assigned automatically and immediately to another, available, node.  
+   Supported tasks include Reads, Writes, and the ongoing tasks SQL and Raven ETL, Backup, Data subscription, 
+   and External replication.  
+   If, for example, the node responsible for a Raven ETL task fails, the cluster observer will assign
```
Comment (karmeli87) [inline Documentation/5.2/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown:36]: same

File: Documentation/5.2/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+   Supported tasks include Reads, Writes, and the ongoing tasks SQL and Raven ETL, Backup, Data subscription, 
+   and External replication.  
+   If, for example, the node responsible for a Raven ETL task fails, the cluster observer will assign 
+   an available node with the responsibility for the task and ETL transfers will automatically resume 
+   from the failure point on.  
+ * If your license does **not** provide highly available tasks, the tasks will resume their activity when 
+   the original node returns online.  
+   If the fallen node does not return online within a given time (set by 
+   [cluster.timebeforeaddingreplicainsec](../../../server/configuration/cluster-configuration#cluster.timebeforeaddingreplicainsec), 
+   the cluster observer will attempt to select an available node to replace it in the database group
```
Comment (karmeli87) [inline Documentation/5.2/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown:43]: this is dynamic database distribution feature.

File: Documentation/5.2/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+* `Responsible Node` is the node that is responsible to perform a specific Ongoing Task.  
+
+* Each node checks whether it is the `Responsible Node` for the task by executing a local function that is based on the  
+  _unique hash value_ of the task and the current [Database Topology](../../../server/clustering/distribution/distributed-database#database-topology).  
+
+* Since the `Database Topology` is _eventually consistent_ across the cluster,  
+  there will be an **eventually consistent single Responsible Node**, which will answer the above constraints.  
+
+{NOTE: Mentor Node}
+The node is called a `Mentor Node` when its task is updating a _Rehab_ or a _Promotable_.
```
Comment (karmeli87) [inline Documentation/5.2/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown:75]: Rehab is just a specific case of the `Mentor Node`, but it applies to all tasks

File: Documentation/5.2/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+The node is called a `Mentor Node` when its task is updating a _Rehab_ or a _Promotable_.  
+{NOTE/}
+{PANEL/}
+
+{PANEL: Tasks Relocation}
+
+* Upon a `Database Topology` change, _all_ existing tasks will be re-evaluated and 
+  re-distributed among the functional nodes.   
+
+* The responsible node for an `Ongoing Task` is also re-evalutated upon a change in the
```
Comment (karmeli87) [inline Documentation/5.2/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown:84]: Not sure I understand what does it mean

File: Documentation/5.2/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+It is sometimes preferable to **prevent** the failover of tasks to different responsible nodes.  
+An example for such a case is a heavy duty backup task, that better be left for the continuous care 
+of its original node than reassigned during a backup operation.  
+Another example is an ETL task that transfers 
+[artificial documents](../../../studio/database/indexes/create-map-reduce-index#saving-map-reduce-results-in-a-collection-(artificial-documents)). 
+In this case a reassigned task might skip some of the artificial documents that were created on 
+the original node.  
+
+The failover of a task to another responsible node can be prevented by **pinning the task** to 
+its original node.
```
Comment (karmeli87) [inline Documentation/5.2/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown:118]: to mentor node

File: Documentation/5.1/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown [hunk 8]
```diff
@@ -15,13 +15,34 @@
...
   * [Constraints](../../../server/clustering/distribution/highly-available-tasks#constraints)  
   * [Responsible Node](../../../server/clustering/distribution/highly-available-tasks#responsible-node)  
   * [Tasks Relocation](../../../server/clustering/distribution/highly-available-tasks#tasks-relocation)  
 {NOTE/}
 
 ---
 
+{PANEL: License}
+
+Please [check your license](https://ravendb.net/buy) to verify whether the Highly Available Tasks feature
```
[the remark this answers is not in the document]
Comment (reebhub) [reply Documentation/5.1/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown:28]: right
done

File: Documentation/5.1/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown [hunk 9]
```diff
@@ -15,13 +15,34 @@
...
 
 ---
 
+{PANEL: License}
+
+Please [check your license](https://ravendb.net/buy) to verify whether the Highly Available Tasks feature 
+is activated in your database.  
+  
+ * If your license **provides** highly available tasks, the responsibilities of a failed node will be assigned 
+   automatically and immediately to another, available, cluster node.
```
[the remark this answers is not in the document]
Comment (reebhub) [reply Documentation/5.1/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown:32]: done

File: Documentation/5.1/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown (hunk 1 above)
[the remark this answers is not in the document]
Comment (reebhub) [reply Documentation/5.1/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown:33]: updated to:
`Supported tasks include Backup, Data subscription, and all ETL and Replication types. `

File: Documentation/5.2/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown (hunk 2 above)
Comment (reebhub) [reply Documentation/5.2/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown:9]: true, we're just using the opportunity to explain our behavior, when we find a better place for it we'll probably move it there.

File: Documentation/5.2/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown (hunk 3 above)
Comment (reebhub) [reply Documentation/5.2/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown:36]: right

File: Documentation/5.1/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown (hunk 1 above)
[the remark this answers is not in the document]
Comment (reebhub) [reply Documentation/5.1/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown:33]: added this issue, to verify & consider what's next about replication
https://issues.hibernatingrhinos.com/issue/RavenDB-22003/Replication-tasks-do-not-run-as-highly-available-tasks

File: Documentation/5.2/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown (hunk 4 above)
Comment (reebhub) [reply Documentation/5.2/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown:43]: ok, removed this part

File: Documentation/5.2/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown (hunk 5 above)
Comment (reebhub) [reply Documentation/5.2/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown:75]: done

File: Documentation/5.2/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown (hunk 7 above)
Comment (reebhub) [reply Documentation/5.2/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown:118]: done

File: Documentation/5.2/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown (hunk 6 above)
Comment (reebhub) [reply Documentation/5.2/Raven.Documentation.Pages/server/clustering/distribution/highly-available-tasks.markdown:84]: --> When the unique hash code of an ongoing task changes, the node responsible for this 
  task will be re-evaluated as well.

**Pull request closed without merging**